### DISCUSSION TITLE
Querying through nested composite primary keys

### DISCUSSION THREAD

**Discussion by AdriVanHoudt** [mikro-orm/mikro-orm#3470] [Q&A]
**Describe the bug**
I'm trying to query through nested composite keys but somehow the output is not what I expect.
I suspect I'm either trying something I'm not supposed to or my entities are not setup correctly.

**To Reproduce**
https://stackblitz.com/edit/mikro-orm-repro-bw48ez?file=test.ts

**Expected behavior**
A join on `entity` with the id and org id.

**Additional context**

**Versions**

| Dependency | Version |
| - | - |
| node | 16 |
| typescript | 4.7 (but latest fails as well) |
| mikro-orm | 5.3.1 (but latest fails as well) |
| your-driver | postgres |

**Comment by B4nan** [comment]
You sent here a stackblitz with a test, but its passing, it does not assert anything. And I dont see any description of what you think is wrong in your issue either. I would also appreciate less cryptic entity names, having property called `entity`, and entities like `EntityEntity`, tests should show the problem, this hides it.

The queries look good to me, I dont think there should be any joins. 

- your `EntityInfo` entity has a composite PK -> both of the parts are known without any joins
- it also has `entity` property - a m:1 relation, so again, its PK is known
- this PK is again composite, we know both its parts
- you query only known properties, hence no joins needed

So the question is, why do you expect a join?

**Reply by AdriVanHoudt** [reply]
Thanks for the quick response!
Sorry about the tests.
The only assert I could've added was to make sure the generated query doesn't include `entity_organization_id` as that column doesn't exist.

I can't do anything about the entity name, I'm working against an existing DB with a table called `entity` and I wanted to make sure that was not a reason for the issue. I can't name the entity `Entity` as it conflicts with the `@Entity` decorator.

I expect a join since I'm asking `give me entity info where entity id is 1 and where the organization id of the entity is 2`.
But if it were to resolve to 
```sql
 select `e0`.* from `entity_info` as `e0` where (`e0`.`entity_id`, `e0`.`organization_id`) in ( values (?, ?))`
```
that would be ok but it resolves to 
```sql
    select `e0`.* from `entity_info` as `e0` where (`e0`.`entity_id`, `e0`.`entity_organization_id`) in ( values (?, ?))
```
now (see the `entity_organization_id` column which does not exist).

I hope this clears things up. Let me know if things are not clear.

**Reply by B4nan** [reply]
> The only assert I could've added was to make sure the generated query doesn't include entity_organization_id as that column doesn't exist.

It does exist, at least that is what your entity definition says. You have a m:1 property called `entity`, which has a composite PK - so it will be represented by two columns, prefixed with `entity_`.

It sounds like you want to have the org column shared between two properties - this is indeed not supported. I dont mind supporting such use case, but at the same time, I have no idea how to support it in the entity definition. It cant be default behaviour, that does not make much sense.

**Reply by AdriVanHoudt** [reply]
Ah that is prob the issue indeed.
The schema in the db is as follows:
organization.id
entity.id, entity.organization_id -> composite PK, with organization_id being a FK to organization.id
entityinfo.id, entityinfo.organization_id -> composite PK , with entityinfo.organization_id, entityinfo.entity_id being a composite FK to entity.id, entity.organization_id

My question would be to how to best model this in mikro orm, I guess there is no current way to keep the relations and all the keys?
So I probably would just remove the organization ManyToOne and just use that column as is (we would never get the org from the entity info directly in practice any way).
Can I use `mapToPk` for that, not super sure what it does?

**Reply by B4nan** [reply]
Nope, there is no workaround for this. You want to share single database column in multiple properties, and that is not supported. It does not matter if you map to entities or to their PKs, still the same issue.

**Reply by B4nan** [reply]
Do you know some other ORMs that do support this? How does the entity definition look like? I think it could be rather simple to make it work, we just need to find a way to tell the ORM what you want (in other words, to tell the ORM that the org column should be part of multiple properties).

Maybe we could just force the field name to be same, it might still require some adjustments in the ORM, but it might be enough.

**Reply by AdriVanHoudt** [reply]
So the problem is that both EntityEntity en EntityInfo have a organization_id in the PK right?

Note that these org ids are always the same. So any EntityEntity linked to EntityInfo have the same org id.

I do not know of an ORM that allows it (not that I have much to reference), mabe hibernate?

**Reply by AdriVanHoudt** [reply]
To clarify the organization_id is basically a tenant id used for data isolation

**Reply by B4nan** [reply]
> So the problem is that both EntityEntity en EntityInfo have a organization_id in the PK right?

The problem is that you want this column to be just one, but you have two properties, each having its own org column. 

> Note that these org ids are always the same. 

Yeah I get that. But from the entity definition you have, its easily possible to have them on different values. I understand that is what you want to "disallow" on the schema level, but the ORM does not allow it currently. 

I think we could make it work if you override the column name explicitly (but again, it will need some changes in the codebase, not a workaround you could use right now). Will need to give it some time and experiment a bit and report back. 

FYI you are not the first one asking for this (maybe second or third), at the same time it seems to be rather edge case so it wasn't prioritized.

**Reply by AdriVanHoudt** [reply]
I understand, thanks for the quick responses!
If you want more input or want something tested/validated let me know.
